Basic support for images in ODT documents

hubertp-lshift opened the pull request:
Highly influenced by the docx support, refactored
some code to avoid DRY.

Review by @akahl-lshift

hubertp committed:
Basic support for images in ODT documents

Highly influenced by the docx support, refactored
some code to avoid DRY.

hubertp-lshift commented:
@akahl-lshift: don't merge I just created a PR for an internal review.

akahl-owl commented on src/Text/Pandoc/Shared.hs:
```diff
@@ -1028,6 +1031,19 @@ collapseFilePath = Posix.joinPath . reverse . foldl go [] . splitDirectories
+filteredFilesFromArchive zf f =
+  mapMaybe (fileAndBinary zf) (filter f (filesInArchive zf))
+  where
+    fileAndBinary :: Archive -> FilePath -> Maybe (FilePath, BL.ByteString)
```
I personally would prefer to have `fileAndBinary` not inlined in order to be able to test it. Since it's Haskell and type checking already does 80% of the job verifying the program will work, decide for yourself.

hubertp-lshift replied:
Ok, I come from a OO world and to me it just feels more like an inner function.

akahl-owl commented on src/Text/Pandoc/Shared.hs:
```diff
@@ -1028,6 +1031,19 @@ collapseFilePath = Posix.joinPath . reverse . foldl go [] . splitDirectories
+  where
+    fileAndBinary :: Archive -> FilePath -> Maybe (FilePath, BL.ByteString)
+    fileAndBinary a fp =
+      case findEntryByPath fp a of
```
I'm no expert but I think it would be idiomatic to use the bind function `>>=` for the Maybe monad, here.
`findEntryByPath fp a >>= \e -> Just (fp, fromEntry e)`

See e.g. http://stackoverflow.com/a/10881861

hubertp-lshift replied:
Thanks

akahl-owl commented on src/Text/Pandoc/Readers/Odt/ContentReader.hs:
```diff
@@ -190,6 +206,22 @@ popStyle =     keepingTheValue (
+           >>^ fst
+
+lookupResource :: OdtReaderSafe String (FilePath, B.ByteString)
+lookupResource = proc target -> do
```
Could you enlighten me why this piece of imperative code is needed in Haskell holy stateless kingdom?

hubertp-lshift replied:
Apparently that's how they carry around state and I didn't want to stand out.

akahl-owl commented on src/Text/Pandoc/Readers/Odt/ContentReader.hs:
```diff
@@ -190,6 +206,22 @@ popStyle =     keepingTheValue (
 
+--
+updateMediaWithResource :: OdtReaderSafe (FilePath, B.ByteString) (FilePath, B.ByteString)
+updateMediaWithResource = keepingTheValue (
```
The indentation looks really odd, is this really intended? Can't most of this ago onto a single line?

hubertp-lshift replied:
See a function a couple of lines above.

akahl-owl commented on src/Text/Pandoc/Readers/Odt/ContentReader.hs:
```diff
@@ -688,6 +722,46 @@ read_table_cell    = matchingElement NsTable "table-cell"
                                           ]
 
 ----------------------
+-- Images
```
I hate myself and want to die reading this code.

hubertp-lshift replied:
Not sure if I can help you with that.

akahl-owl reviewed (approved):
I'd say it's OK. If I understand correctly, the imperative portions are a consequence of the external libs you're using, right?

hubertp committed:
Infer caption from the text following the img

Frame can contain other frames with the text boxes.
This is something that has not been considered before
and meant that the whole construction of images was
broken in those cases. Also the captions were fixed/ignored.

hubertp committed:
Minor refactoring

hubertp committed:
Use bind function instead of pattern matching

hubertp-lshift commented:
Ready for push to origin.
